src/main/scala/org/kframework/minikore/MiniKore.scala

```diff
@@ -1,39 +1,162 @@
 package org.kframework.minikore
 
+
+import org.kframework.minikore.TreeInterface._
+import org.kframework.minikore.{PatternInterface => i}
+
 import scala.collection._
 
 /** Algebraic data type of MiniKore. */
 object MiniKore {
 
-  type Attributes = Seq[Pattern]
+  //Default Implementation
+
+  type Attributes = Seq[i.Pattern]
 
   case class Definition(modules: Seq[Module], att: Attributes)
+
   case class Module(name: String, sentences: Seq[Sentence], att: Attributes)
 
   sealed trait Sentence
+
   case class Import(name: String, att: Attributes) extends Sentence
+
   case class SortDeclaration(sort: String, att: Attributes) extends Sentence
+
   case class SymbolDeclaration(sort: String, label: String, args: Seq[String], att: Attributes) extends Sentence
-  case class Rule(pattern: Pattern, att: Attributes) extends Sentence
-  case class Axiom(pattern: Pattern, att: Attributes) extends Sentence
-
-  sealed trait Pattern
-  case class Variable(name: String, sort: String) extends Pattern
-  case class Application(label: String, args: Seq[Pattern]) extends Pattern
-  case class DomainValue(label: String, value: String) extends Pattern
-  //
-  case class True() extends Pattern
-  case class False() extends Pattern
-  //
-  case class And(p: Pattern, q: Pattern) extends Pattern
-  case class Or(p: Pattern, q: Pattern) extends Pattern
-  case class Not(p: Pattern) extends Pattern
-  case class Implies(p: Pattern, q: Pattern) extends Pattern
-  case class Exists(v: Variable, p: Pattern) extends Pattern
-  case class ForAll(v: Variable, p: Pattern) extends Pattern
-  //
-  case class Next(p: Pattern) extends Pattern
-  case class Rewrite(p: Pattern, q: Pattern) extends Pattern
-  case class Equal(p: Pattern, q: Pattern) extends Pattern
+
+  case class Rule(pattern: i.Pattern, att: Attributes) extends Sentence
+
+  case class Axiom(pattern: i.Pattern, att: Attributes) extends Sentence
+
+  case class Variable(name: String, sort: i.Sort) extends i.Variable {
+    override def build = DefaultBuilders.VariableBuilder
+  }
+
+  case class Application(label: String, args: Seq[i.Pattern]) extends i.Application {
+    override def build = DefaultBuilders.ApplicationBuilder(label, _)
+  }
+
+
+  case class DomainValue(label: String, value: String) extends i.DomainValue {
+    override def build = DefaultBuilders.DomainValueBuilder
+  }
+
+  case class True() extends i.True {
```

Comment:
No, they are compiled differently. To use the case class one would have to write `val x = True()` which would create a new, identical instance of the case class at each invocation. Using `object` would create a singleton, with the usual property that only one instance would ever be created, and that instance would be available by calling `True`. So it would be: `val x = True`.

My preference is for using object as it encodes the constant nature of True at the language level. Also, I think it should be called `Top` to distinguish it from the `True` in the defined language.